### ISSUE TITLE
Fix Lint/Void false positive with macro hash arguments.

### ISSUE THREAD

**Issue by bkuhlmann** [rubocop-hq/rubocop#6440]
## Expected behavior

When writing [Thor](http://whatisthor.com)-based CLI programs, Rubocop flags the use of the `map` macro due to the hash argument (I believe). Example:

```
map %w[-t --test] => :test
```

This does not happen when passing a string to Thor's `desc` macro. Example:

```
desc "-t, [--test=MESSAGE]", "Print test message."
```

I would expect/think Rubocop to consider the use of `map %w[-t --test] => :test` to valid.

## Actual behavior

When running Rubocop on the example source code, I see the following error:

```
example.rb:22:5: W: Lint/Void: Method #map used in void context. Did you mean #map!?
    map %w[-t --test] => :test
    ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

## Steps to reproduce the problem

You can reproduce this error by saving the following code (save it as `example.rb`) and running it locally:

```
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError
  abort "Bundler 1.10 or later is required."
end

gemfile true do
  source "https://rubygems.org"
  gem "thor", "~> 0.20.0"
end

require "thor"
require "thor/actions"

module Example
  class CLI < Thor
    include Thor::Actions

    desc "-t, [--test=MESSAGE]", "Print test message."
    map %w[-t --test] => :test
    def test message
      say "This is a test: #{message}."
    end
  end
end

Example::CLI.start
```

Usage:

```
# Run the program (optional).
ruby example.rb -t "Hi"

# Run Rubocop
rubocop --display-cop-names example.rb
```

## RuboCop version

```
> rubocop -V
0.60.0 (using Parser 2.5.3.0, running on ruby 2.5.3 x86_64-darwin17)
```

**Comment by rrosenblum** [comment]
This is one of the problems with static code analysis. It is nearly impossible for us to tell the difference in `Thor#map` and `Enumerable#map`. With `Enumerable#map` most of the time, there will be an object that receives this method call. This is not the case when invoking `map` in a class that extends `Enumerable`. 

The question becomes which is the better edge case to account for? Should we ignore calls to `map` that don't have a receiver when checking this cop? Are we better off leaving the current implementation and having the user disable this cop for files that make use of `Thor`?

This is only reproducible when `CheckForMethodsWithNoSideEffects` is set to `true`.

**Comment by rrosenblum** [comment]
One improvement that we might be able to make to this cop is to check the arguments that are being passed to `map`. `Enumerable#map` will be invoked with a symbol proc or a block. If it is invoked with something else, we can probably assume that isn't the implementation from `Enumerable#map`.

**Comment by bkuhlmann** [comment]
> Are we better off leaving the current implementation and having the user disable this cop for files that make use of Thor?

Based on your feedback, I'm thinking current implementation might be good enough.

> This is only reproducible when CheckForMethodsWithNoSideEffects is set to true.

Thanks, yeah, forgot to mention that. 😅 

> One improvement that we might be able to make to this cop is to check the arguments that are being passed to `map`.

I don't know. There's edge cases there too. You might also want to get the enumerator back for lazy execution in the future (i.e. no args).

I'll leave it up to you to close this issue. As a workaround I can exclude source that uses Thor, switch to a different CLI tool, or use inline comments.
